### PR TITLE
SI-9760 Fix for higher-kinded GADT refinement

### PR DESCRIPTION
**Pull request by milessabin**

### CODE REVIEW

File: test/files/pos/hkgadt.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+  def frob[F[_]](foo: Foo[F]): F[Int] =
+    foo match {
+      case Bar() =>
+         List(1)
+    }
+
+  sealed trait Foo1[F]
+  final case class Bar1() extends Foo1[Int]
+  def frob1[A](foo: Foo1[A]) = foo match {
```
Comment (larsrh) [inline test/files/pos/hkgadt.scala]: Are you missing a return type here?

Comment (milessabin) [inline test/files/pos/hkgadt.scala]: Yes. It should be `A`. Will fix.

**Comment by adriaanm** [comment]
FTR, this condition was added as a workaround for https://issues.scala-lang.org/browse/SI-4070. Since its tests are still passing, I guess we've improved existentials enough by now that they can kind of quantify over type constructors...

Could you also play around with some neg test cases? Maybe something with more than one case to exercise lubbing?

**Comment by milessabin** [comment]
There's more information on the [ticket](https://issues.scala-lang.org/browse/SI-9760). At @retronym's suggestion I git-bisected my way to determinng that [this commit of yours](https://github.com/scala/scala/pull/2354/commits/6a61e172d9de912c74af50e49f0057fbe8d4e500) was what made the additional condition unnecessary.

I'll add some more tests.

**Comment by milessabin** [comment]
@adriaanm more tests (multiple cases for pos, and some neg cases) added.

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**